src/skin/qml/qmlcontrolproxy.cpp

```diff
@@ -0,0 +1,157 @@
+#include "skin/qml/qmlcontrolproxy.h"
+
+#include "moc_qmlcontrolproxy.cpp"
+
+namespace mixxx {
+namespace skin {
+namespace qml {
+
+QmlControlProxy::QmlControlProxy(QObject* parent)
+        : QObject(parent),
+          m_pControlProxy(nullptr) {
+}
+
+bool QmlControlProxy::isKeyValid() const {
+    return m_coKey.isValid();
+}
+
+bool QmlControlProxy::isInitialized() const {
+    return m_pControlProxy != nullptr;
+}
+
+void QmlControlProxy::setGroup(const QString& group) {
+    const bool keyValidBeforeSet = isKeyValid();
+    m_coKey.group = group;
+    const bool keyValidAfterSet = isKeyValid();
+    if (keyValidBeforeSet != keyValidAfterSet) {
+        emit keyValidChanged(keyValidAfterSet);
+    }
+    tryInitialize();
+}
+
+const QString& QmlControlProxy::getGroup() const {
+    return m_coKey.group;
+}
+
+void QmlControlProxy::setKey(const QString& key) {
+    const bool keyValidBeforeSet = isKeyValid();
+    m_coKey.item = key;
+    const bool keyValidAfterSet = isKeyValid();
+    if (keyValidBeforeSet != keyValidAfterSet) {
+        emit keyValidChanged(keyValidAfterSet);
+    }
+    tryInitialize();
+}
+
+const QString& QmlControlProxy::getKey() const {
+    return m_coKey.item;
+}
+
+void QmlControlProxy::setValue(double newValue) {
+    if (!isInitialized()) {
+        qWarning() << "QmlControlProxy: Attempted to set value" << newValue
+                   << "on non-initialized CO" << m_coKey;
+        return;
+    }
+    m_pControlProxy->set(newValue);
+    m_pControlProxy->emitValueChanged();
+}
+
+double QmlControlProxy::getValue() const {
+    if (!isInitialized()) {
+        qWarning() << "QmlControlProxy: Attempted to get value from "
+                      "non-initialized CO"
+                   << m_coKey << "(returning 0)";
+        return 0;
+    }
+    return m_pControlProxy->get();
+}
+
+void QmlControlProxy::setParameter(double newValue) {
+    if (!isInitialized()) {
+        qWarning() << "QmlControlProxy: Attempted to set parameter" << newValue
+                   << "on non-initialized CO" << m_coKey;
+        return;
+    }
+    m_pControlProxy->setParameter(newValue);
+    m_pControlProxy->emitValueChanged();
```

Comment:
emitValueChanged() or emitParameterChanged()?

Why do we need keep track of parameter changes? Aren't parameter not just mapped to a value internally?

And why do we emit value changes separately when set explicitly instead directly connecting the valueChanged() signal of ControlProxy?

Did I misunderstand the purpose of QmlControlProxy and its relation to the wrapped ControlProxy?

Reply:
Following problem: A control proxy doesn't receive signals if it's the origin of the value change. This means the following will not work if we don't emit these signals:

````qml
Item {
    Mixxx.ControlProxy {
        id: mycontrol
        group: "[Channel1]"
        key: "play"
    }

    Button {
         onClicked: mycontrol.value = (mycontrol.value) == 0 ? mycontrol.value = 1 : 0
    }

    Rectangle {
        color: "red"
        opacity: mycontrol.parameter
    }
}
```

However, this should possibly be handled in the QmlControlProxy itself, emit signals on the underlying control proxy.